massa-consensus-worker/src/state/prune.rs

```diff
@@ -135,30 +137,41 @@ impl ConsensusState {
         });
     }
 
-    // Keep only a certain (`config.max_future_processing_blocks`) number of blocks that are discarded
+    // Keep only a certain (`config.max_discarded_blocks`) number of blocks that are discarded
     // to avoid too big memory consumption
     fn prune_discarded(&mut self) -> Result<(), ConsensusError> {
         if self.discarded_index.len() <= self.config.max_discarded_blocks {
             return Ok(());
         }
-        let mut discard_hashes: Vec<(u64, BlockId)> = self
+        let mut discard_hashes: Vec<(u64, Slot, BlockId)> = self
             .discarded_index
             .iter()
             .filter_map(|block_id| {
                 if let Some(BlockStatus::Discarded {
-                    sequence_number, ..
+                    sequence_number,
+                    slot,
+                    ..
                 }) = self.block_statuses.get(block_id)
                 {
-                    return Some((*sequence_number, *block_id));
+                    return Some((*sequence_number, *slot, *block_id));
                 }
                 None
             })
             .collect();
         discard_hashes.sort_unstable();
         discard_hashes.truncate(self.discarded_index.len() - self.config.max_discarded_blocks);
-        for (_, block_id) in discard_hashes.iter() {
+        for (_, slot, block_id) in discard_hashes.iter() {
             self.block_statuses.remove(block_id);
             self.discarded_index.remove(block_id);
+            match self.blocks_per_slot.entry(*slot) {
```

Comment:
Instead of clearing there, we should remove all slots in blocks_per_slot at or below the latest final slot of every thread.

Reply:
Done.